File: lovebabbar/linked_list/linked_list.cpp
#include <bits/stdc++.h>

using namespace std;

class Node {
   public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = NULL;
    }
};

class LinkedList {
   public:
    Node* head;

    LinkedList() {
        head = NULL;
    }

    Node* createNode(int data) {
        return new Node
    }

    void insertLast(int data) {
        Node* newNode = new Node(data);

        if (head == NULL) {
            head = newNode;
            return;
        }

        Node* top = head;
        while (top->next != NULL) top = top->next;

        top->next = newNode;
    }

    void insertFirst(int data) {
        Node* newNode = new Node(data);

        if (head == NULL) {
            head = newNode;
            return;
        }

        Node* top = head;
        newNode->next = top;
        head = newNode;

        return;
    }

    void insertAfter(int data, int K) {
        Node* newNode = new Node(data);

        if (head == NULL) {
            head = newNode;
            return;
        }

        if (K == 0) {
            insertFirst(data);
            return;
        }

        Node* top = head;
        int cnt = K - 1;
        while (top->next != NULL && cnt--) {
            top = top->next;
        }

        if (top == NULL) {
            cout << "Enter valid value\n";
            return;
        }

        newNode->next = top->next;
        top->next = newNode;
        return;
    }

    void removeLast() {
        if (head == NULL) {
            cout << "There is no node in the linked list!\n";
            return;
        }

        if (head->next == NULL) {
            head = NULL;
            return;
        }

        Node* top = head;
        while (top->next->next != NULL) {
            top = top->next;
        }

        top->next = NULL;
    }

    void removeFirst() {
        if (head == NULL) {
            cout << "There is no node in the linked list!\n";
            return;
        }

        if (head->next == NULL) {
            head = NULL;
            return;
        }

        head = head->next;
        return;
    }

    void removeAfter(int K) {
        if (head == NULL) {
            cout << "Linked List is empty!";
            return;
        }

        if (K == 0) {
            removeFirst();
            return;
        }

        Node* top = head;
        int cnt = K - 1;
        while (top->next->next != NULL && cnt--) {
            top = top->next;
        }

        if (top == NULL || cnt>0) {
            cout << "Enter valid value\n";
            return;
        }

        Node* temp = top->next->next;

        top->next = temp;
        return;
    }

    void printList() {
        Node* top = head;

        cout << "\n";

        while (top != NULL) {
            cout << top->data << " ";
            top = top->next;
        }

        cout << "\n\n";
        return;
    }
};

// int main() {
//     return 0;
// }

File: Java/src/LinkedLists/LinkedList.java
package LinkedLists;
import java.io.*;

public class LinkedList {

    Node head;
    
    static class Node {

        int data;
        Node next;

        Node(int d)
        {
            data = d;
            next = null;
        }
    }
    
    int size() {
        Node ptr = head;
        int cnt = 0;
        
        while(ptr!=null) {
            ptr = ptr.next;
            cnt++;
        }
        return cnt;
    }

    public LinkedList insertAtLast(LinkedList list, int data)
    {
        Node new_node = new Node(data);
        
        if (list.head == null) {
            list.head = new_node;
        }
        else {
            Node last = list.head;
            while (last.next != null) {
                last = last.next;
            }

            last.next = new_node;
        }

        return list;
    }
    
    public LinkedList insertAtFirst(LinkedList list,  int data) {
        Node new_node = new Node(data);

        if (list.head == null) {
            list.head = new_node;
        }
        else {
            new_node.next = list.head;
            list.head = new_node;
        }

        return list;
    }

    public LinkedList insertAfterNthNode(LinkedList list, int data, int position) {
        Node new_node = new Node(data);

        if (list.head == null) {
            list.head = new_node;
        }
        else if(position==0) {
            insertAtFirst(list, data);
        }
        else if(position == list.size()) {
            insertAtLast(list, data);
        }
        else {
            
            Node ptr = list.head;
            int cnt = position-1;
            while(cnt > 0 && ptr != null) {
                ptr = ptr.next;
                cnt--;
            }
            
            new_node.next = ptr.next;
            ptr.next = new_node;
        }

        return list;
    }

    public LinkedList deleteLast(LinkedList list) {
        if(list.head == null) return list;
        else if(list.head.next == null) {
            list.head = null;
            return list;
        }
        else {
            Node ptr = list.head;
            
            while(ptr.next.next!=null) {
                ptr = ptr.next;
            }
            
            ptr.next = null;
            return list;
        }
    }

    public LinkedList deleteFirst(LinkedList list) {
        if(list.head == null) return list;
        else if(list.head.next == null) {
            list.head = null;
            return list;
        }
        else {
            list.head = list.head.next;
            return list;
        }
    }

    public LinkedList deleteNthNode(LinkedList list, int position) {
        if(list.head == null) return list;
        else if(position == 0) {
            list.head = null;
            return list;
        }
        else {
            list.head = list.head.next;
            return list;
        }
    }
    
    public void printList(LinkedList list)
    {
        Node currNode = list.head;

        System.out.print("LinkedList: ");

        while (currNode != null) {
            System.out.print(currNode.data + " ");
            currNode = currNode.next;
        }
    }

    public static void main(String[] args)
    {
        LinkedList list = new LinkedList();

        list = list.insertAtLast(list, 1);
        list = list.insertAtLast(list, 2);
//        list = insertAtLast(list, 3);
//        list = deleteFirst(list);
        list = list.deleteNthNode(list, 2);
        
        list.printList(list);
    }
}
